lock-api/src/main/java/com/palantir/lock/client/BlockEnforcingLockService.java

```diff
@@ -0,0 +1,168 @@
+import com.palantir.atlasdb.timelock.api.ConjureLockResponse;
+import com.palantir.atlasdb.timelock.api.SuccessfulLockResponse;
+import com.palantir.atlasdb.timelock.api.UnsuccessfulLockResponse;
+import com.palantir.lock.v2.LockRequest;
+import com.palantir.lock.v2.LockResponse;
+import com.palantir.lock.v2.WaitForLocksRequest;
+import com.palantir.lock.v2.WaitForLocksResponse;
+import com.palantir.logsafe.Preconditions;
+import com.palantir.logsafe.SafeArg;
+import com.palantir.logsafe.exceptions.SafeIllegalStateException;
+
+/**
+ * Ensures that clients actually attempt to acquire the lock for the full duration they claim they will block for,
+ * unless they run into an exception that is unlikely to actually be a timeout. This is done to account for the reality
+ * of bounded timeouts beneath us (e.g. in terms of the networking layer) when communicating with remote services.
+ *
+ * Fairness is admittedly compromised, but this is a closer approximation than the previous behaviour.
+ */
+final class BlockEnforcingLockService {
+    private final NamespacedConjureTimelockService namespacedConjureTimelockService;
+    private final RemoteTimeoutRetryer timeoutRetryer;
+
+    private BlockEnforcingLockService(NamespacedConjureTimelockService namespacedConjureTimelockService,
+            RemoteTimeoutRetryer timeoutRetryer) {
+        this.namespacedConjureTimelockService = namespacedConjureTimelockService;
+        this.timeoutRetryer = timeoutRetryer;
+    }
+
+    static BlockEnforcingLockService create(NamespacedConjureTimelockService namespacedConjureTimelockService) {
+        return new BlockEnforcingLockService(namespacedConjureTimelockService, RemoteTimeoutRetryer.createDefault());
+    }
+
+    LockResponse lock(LockRequest request) {
+        // The addition of a UUID takes place only at the Conjure level, so we must retry the same request.
+        return timeoutRetryer.attemptUntilTimeLimitOrException(
+                ConjureLockRequests.toConjure(request),
+                Duration.ofMillis(request.getAcquireTimeoutMs()),
+                BlockEnforcingLockService::clampLockRequestToDeadline,
+                this::performSingleLockRequest,
+                response -> !response.wasSuccessful());
+    }
+
+    WaitForLocksResponse waitForLocks(WaitForLocksRequest request) {
+        return timeoutRetryer.attemptUntilTimeLimitOrException(
+                ConjureLockRequests.toConjure(request),
+                Duration.ofMillis(request.getAcquireTimeoutMs()),
+                BlockEnforcingLockService::clampLockRequestToDeadline,
+                this::performSingleWaitForLocksRequest,
+                response -> !response.wasSuccessful());
+    }
+
+    private static ConjureLockRequest clampLockRequestToDeadline(ConjureLockRequest request, Duration remainingTime) {
+        return ConjureLockRequest.builder()
+                .from(request)
+                .acquireTimeoutMs(Ints.saturatedCast(remainingTime.toMillis()))
```

Comment:
can we Ints.checkedCast this? This is only likely to be a long if we have a bug.

Reply:
Sure, yeah, agree `checkedCast` is more appropriate.